cpp/src/tsne/tsne.cu

```diff
@@ -96,70 +102,183 @@ void TSNE_fit(const cumlHandle &handle, const float *X, float *Y, const int n,
+  }
+  else {
+    A = X;
+  }
+
   START_TIMER;
   //---------------------------------------------------
   // Get distances
   if (verbose) printf("[Info] Getting distances.\n");
-  float *distances =
-    (float *)d_alloc->allocate(sizeof(float) * n * n_neighbors, stream);
-  long *indices =
-    (long *)d_alloc->allocate(sizeof(long) * n * n_neighbors, stream);
-  TSNE::get_distances(X, n, p, indices, distances, n_neighbors, stream);
+
+  device_buffer<float> distances_(d_alloc, stream, n*n_neighbors);
+  float *distances = distances_.data();
+  device_buffer<long> indices_(d_alloc, stream, n*n_neighbors);
+  long *indices = indices_.data();
+  TSNE::get_distances(A, n, p, indices, distances, n_neighbors, stream);
+
+  if (pca_intialization == true) {
+    X_C_contiguous.resize(0, stream);  // remove C contiguous layout
+  }
+
   //---------------------------------------------------
   END_TIMER(DistancesTime);
 
   START_TIMER;
   //---------------------------------------------------
   // Normalize distances
-  if (verbose)
+  if (verbose) {
     printf("[Info] Now normalizing distances so exp(D) doesn't explode.\n");
+  }
   TSNE::normalize_distances(n, distances, n_neighbors, stream);
   //---------------------------------------------------
   END_TIMER(NormalizeTime);
 
   START_TIMER;
   //---------------------------------------------------
   // Optimal perplexity
-  if (verbose)
+  if (verbose) {
     printf("[Info] Searching for optimal perplexity via bisection search.\n");
-  float *P =
-    (float *)d_alloc->allocate(sizeof(float) * n * n_neighbors, stream);
-  const float P_sum =
-    TSNE::perplexity_search(distances, P, perplexity, perplexity_max_iter,
-                            perplexity_tol, n, n_neighbors, handle);
-  d_alloc->deallocate(distances, sizeof(float) * n * n_neighbors, stream);
-  if (verbose) printf("[Info] Perplexity sum = %f\n", P_sum);
+  }
+
+  int SAVED_SPACE = 0;
+  const int NNZ = (2 * n * n_neighbors);
+
+  device_buffer<float> P_(d_alloc, stream, NNZ);
+  SAVED_SPACE += n*n_neighbors*sizeof(float);
+
+  float *P = P_.data();
+  TSNE::perplexity_search(distances, P, perplexity, perplexity_max_iter,
+                          perplexity_tol, n, n_neighbors, handle);
+
+  distances_.resize(0, stream);
   //---------------------------------------------------
   END_TIMER(PerplexityTime);
 
+
   START_TIMER;
   //---------------------------------------------------
   // Convert data to COO layout
-  MLCommon::Sparse::COO<float> COO_Matrix;
-  TSNE::symmetrize_perplexity(P, indices, n, n_neighbors, P_sum,
-                              early_exaggeration, &COO_Matrix, stream, handle);
-  const int NNZ = COO_Matrix.nnz;
-  float *VAL = COO_Matrix.vals;
-  const int *COL = COO_Matrix.cols;
-  const int *ROW = COO_Matrix.rows;
+  // MLCommon::Sparse::COO<float> COO_Matrix;
+  
+  float *VAL = P;
+  device_buffer<int> COL_(d_alloc, stream, NNZ);
+  int *COL = COL_.data();
+
+  int *ROW = (int*)indices;
```

Comment:
Is your purpose for this conversion, and the next few lines,  to make the code portable to compilers that use 32-bits for longs? I would use int64_t here to be concise. I don’t believe 64-bit long array can be explicitly type casted to 32-bit int arrays like this.

Reply:
Ohhh this was actually used to save memory. Originally FAISS provides `long *indices`, but that also uses 2x `int *indices` of memory. So I typecasted it to save memory, since TSNE uses a lot of memory, and so the maximum anyways it can handle is like 30 million points or something. By doing this, I'm trying to reduce memory usage (this PR reduces mem usage by 50%)

Reply:
So `long *indices` is first coverted to `int *COL` in the symmetrization step, then `long *indices` is immediately reused for the `int *ROW` pointers.